adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+    
+    lock_2:
+        \* let mut locked_wait_map = self.wait_map.lock().await;
+        call lock(wait_map_lock);
+    
+    lock_3:
+        \* drop(locked_cache);
+        call unlock(cache_lock);
+    
+    lookup_wait_map:
+        \* match locked_wait_map.get_mut(&key) {
+        \*    Some(waiter) => {
+        if (Cardinality(wait_map[Key]) > 0)
+        {
+            wait_map_subscribe:
+                \* we register to reeive a notification when the value is available
+                \* let mut receiver = waiter.subscribe();
+                wait_map[Key] := wait_map[Key] \union {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+            
+            waiting_for_value:
+                \* let (recv_key, recv_value) = receiver.recv().await
+                await channels[self];
+                goto finished;
+    
+        }
+        else
+        {
+            wait_map_register:
+                \* the wait map should be emty for this key
+                assert wait_map[Key] = {};
+                \* we should not have a value yet
+                assert ~cache[Key];
+                
+                \* let (tx, _rx) = broadcast::channel(1);
+                \* locked_wait_map.insert(key.clone(), tx.clone());
+                wait_map[Key] := {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+                
+            fetching_value:
+                \* let value = self.resolver.retrieve(key.clone()).await;
+                skip;
+            
+            inserting_value_lock_cache:
+                \* let mut locked_cache = self.cached.lock().await;
+                call lock(cache_lock);
+                
+            inserting_value:
+                \* locked_cache.put(key.clone(), value.clone());
+                cache[Key] := TRUE;
+            
+                \* let mut locked_wait_map = self.wait_map.lock().await;
+                call lock(wait_map_lock);
+ 
+            broadcast:
+                \* locked_wait_map.remove(&key);
+                wait_map[Key] := wait_map[Key] \ {self};
+                
+                \* notify all waiting processes
+                \*  tokio::task::spawn_blocking(move || {
+                \*    tx.send((key, broadcast_value))
+                \*        .expect("there is always at least one receiver alive, the _rx guard; qed")
+                \*})
+            notify_all:
+                while(Cardinality(wait_map[Key]) > 0) {
+
+                    pr:= CHOOSE p \in wait_map[Key]: TRUE;
+                    channels[pr] := TRUE;
+                    wait_map[Key] := wait_map[Key] \ {pr};
+                };
+            
+                \* locks are automatically dropped
+                call unlock(wait_map_lock);
+            unlock_1:
+                call unlock(cache_lock);
+            unlock_2:    
+                goto finished;
```

Comment:
the work is done, unlock the wait map and cache. In the rust code, those steps are implicit, they are performed when dropping the mutex guards at the end of the scope